gcc/g++ warnings

Hey,
This seems like an awsome library. I would like to integrate this library to my C/C++ projects, but I'm a gcc/g++ user and the current build presents a few errors/warnings. I don't know if you care to support these compilers. I would love if you did :)

Cheers!

Well, this is my test program:

```
#include "/home/davidoro/libs/PDFGen-master/pdfgen.h"
#include "/home/davidoro/libs/PDFGen-master/pdfgen.c"

int main(int argc, char* argv[]) {
    return 0;
}
```

And this is the output when I run `g++ ./main.cpp -o test`

```
In file included from ./main.cpp:3:
/home/davidoro/libs/PDFGen-master/pdfgen.c:101: warning: "_POSIX_SOURCE" redefined
 #define _POSIX_SOURCE     /* For localtime_r */
 
In file included from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33,
                 from /usr/include/stdint.h:26,
                 from /usr/lib/gcc/x86_64-linux-gnu/8/include/stdint.h:9,
                 from /home/davidoro/libs/PDFGen-master/pdfgen.h:16,
                 from ./main.cpp:2:
/usr/include/features.h:264: note: this is the location of the previous definition
 # define _POSIX_SOURCE 1
 
In file included from ./main.cpp:3:
/home/davidoro/libs/PDFGen-master/pdfgen.c:102: warning: "_XOPEN_SOURCE" redefined
 #define _XOPEN_SOURCE 600 /* for M_SQRT2 */
 
In file included from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33,
                 from /usr/include/stdint.h:26,
                 from /usr/lib/gcc/x86_64-linux-gnu/8/include/stdint.h:9,
                 from /home/davidoro/libs/PDFGen-master/pdfgen.h:16,
                 from ./main.cpp:2:
/usr/include/features.h:203: note: this is the location of the previous definition
 # define _XOPEN_SOURCE 700
 
In file included from ./main.cpp:3:
/home/davidoro/libs/PDFGen-master/pdfgen.c: In function ‘int flexarray_set(flexarray*, int, void*)’:
/home/davidoro/libs/PDFGen-master/pdfgen.c:298:22: error: invalid conversion from ‘void*’ to ‘void***’ [-fpermissive]
         flex->bins = bins;
                      ^~~~
/home/davidoro/libs/PDFGen-master/pdfgen.c:300:19: error: invalid conversion from ‘void*’ to ‘void**’ [-fpermissive]
             calloc(flexarray_get_bin_size(flex, flex->bin_count - 1),
             ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                    sizeof(void *));
                    ~~~~~~~~~~~~~~~
/home/davidoro/libs/PDFGen-master/pdfgen.c: In function ‘ssize_t dstr_ensure(dstr*, size_t)’:
/home/davidoro/libs/PDFGen-master/pdfgen.c:362:27: error: invalid conversion from ‘void*’ to ‘char*’ [-fpermissive]
         new_data = realloc(str->data, new_len);
                    ~~~~~~~^~~~~~~~~~~~~~~~~~~~
/home/davidoro/libs/PDFGen-master/pdfgen.c: In function ‘pdf_object* pdf_get_object(const pdf_doc*, int)’:
/home/davidoro/libs/PDFGen-master/pdfgen.c:478:25: error: invalid conversion from ‘void*’ to ‘pdf_object*’ [-fpermissive]
     return flexarray_get(&pdf->objects, index);
            ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/home/davidoro/libs/PDFGen-master/pdfgen.c: In function ‘pdf_object* pdf_add_object(pdf_doc*, int)’:
/home/davidoro/libs/PDFGen-master/pdfgen.c:525:17: error: invalid conversion from ‘void*’ to ‘pdf_object*’ [-fpermissive]
     obj = calloc(1, sizeof(*obj));
           ~~~~~~^~~~~~~~~~~~~~~~~
/home/davidoro/libs/PDFGen-master/pdfgen.c: In function ‘pdf_doc* pdf_create(float, float, const pdf_info*)’:
/home/davidoro/libs/PDFGen-master/pdfgen.c:576:17: error: invalid conversion from ‘void*’ to ‘pdf_doc*’ [-fpermissive]
     pdf = calloc(1, sizeof(*pdf));
           ~~~~~~^~~~~~~~~~~~~~~~~
/home/davidoro/libs/PDFGen-master/pdfgen.c:591:23: error: invalid conversion from ‘void*’ to ‘pdf_info*’ [-fpermissive]
     obj->info = calloc(sizeof(*obj->info), 1);
                 ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
/home/davidoro/libs/PDFGen-master/pdfgen.c: In function ‘int pdf_save_object(pdf_doc*, FILE*, int)’:
/home/davidoro/libs/PDFGen-master/pdfgen.c:805:30: error: invalid conversion from ‘void*’ to ‘pdf_object*’ [-fpermissive]
                 flexarray_get(&object->page.children, i);
                 ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/davidoro/libs/PDFGen-master/pdfgen.c:834:30: error: invalid conversion from ‘void*’ to ‘pdf_object*’ [-fpermissive]
             f = flexarray_get(&object->bookmark.children, 0);
                 ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/davidoro/libs/PDFGen-master/pdfgen.c:835:30: error: invalid conversion from ‘void*’ to ‘pdf_object*’ [-fpermissive]
             l = flexarray_get(&object->bookmark.children, nchildren - 1);
                 ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/davidoro/libs/PDFGen-master/pdfgen.c: In function ‘pdf_object* pdf_add_raw_jpeg(pdf_doc*, const char*)’:
/home/davidoro/libs/PDFGen-master/pdfgen.c:2239:23: error: invalid conversion from ‘void*’ to ‘uint8_t*’ {aka ‘unsigned char*’} [-fpermissive]
     jpeg_data = malloc(len);
                 ~~~~~~^~~~~
/home/davidoro/libs/PDFGen-master/pdfgen.c: In function ‘int pdf_add_ppm(pdf_doc*, pdf_object*, float, float, float, float, const char*)’:
/home/davidoro/libs/PDFGen-master/pdfgen.c:2339:18: error: invalid conversion from ‘void*’ to ‘uint8_t*’ {aka ‘unsigned char*’} [-fpermissive]
     data = malloc(size);
            ~~~~~~^~~~~~

```

--- a/Makefile
+++ b/Makefile
@@ -33,6 +33,7 @@ tests/penguin.c: data/penguin.jpg
 
 check: testprog pdfgen.c pdfgen.h example-check
 	cppcheck --std=c99 --enable=style,warning,performance,portability,unusedFunction --quiet pdfgen.c pdfgen.h tests/main.c
+	$(CXX) -c pdfgen.c $(CFLAGS_OBJECT) /dev/null -Werror -Wall -Wextra
 	./tests.sh
 	./tests.sh acroread
 	$(CLANG_FORMAT) pdfgen.c | colordiff -u pdfgen.c -

--- a/pdfgen.c
+++ b/pdfgen.c
@@ -100,14 +100,14 @@ typedef SSIZE_T ssize_t;
 #else
 
 #ifndef _POSIX_SOURCE
-#define _POSIX_SOURCE     /* For localtime_r */
+#define _POSIX_SOURCE /* For localtime_r */
 #endif
 
 #ifndef _XOPEN_SOURCE
 #define _XOPEN_SOURCE 600 /* for M_SQRT2 */
 #endif
 
-#include <sys/types.h>    /* for ssize_t */
+#include <sys/types.h> /* for ssize_t */
 #endif
 
 #include <ctype.h>
@@ -297,15 +297,15 @@ static int flexarray_set(struct flexarray *flex, int index, void *data)
     if (bin < 0)
         return -EINVAL;
     if (bin >= flex->bin_count) {
-        void ***bins =
-            (void ***)realloc(flex->bins, (flex->bin_count + 1) * sizeof(flex->bins));
+        void ***bins = (void ***)realloc(flex->bins, (flex->bin_count + 1) *
+                                                         sizeof(flex->bins));
         if (!bins)
             return -ENOMEM;
         flex->bin_count++;
         flex->bins = bins;
         flex->bins[flex->bin_count - 1] =
             (void **)calloc(flexarray_get_bin_size(flex, flex->bin_count - 1),
-                   sizeof(void *));
+                            sizeof(void *));
         if (!flex->bins[flex->bin_count - 1]) {
             flex->bin_count--;
             return -ENOMEM;
@@ -838,8 +838,10 @@ static int pdf_save_object(struct pdf_doc *pdf, FILE *fp, int index)
         int nchildren = flexarray_size(&object->bookmark.children);
         if (nchildren > 0) {
             struct pdf_object *f, *l;
-            f = (struct pdf_object *)flexarray_get(&object->bookmark.children, 0);
-            l = (struct pdf_object *)flexarray_get(&object->bookmark.children, nchildren - 1);
+            f = (struct pdf_object *)flexarray_get(&object->bookmark.children,
+                                                   0);
+            l = (struct pdf_object *)flexarray_get(&object->bookmark.children,
+                                                   nchildren - 1);
             fprintf(fp, "/First %d 0 R\r\n", f->index);
             fprintf(fp, "/Last %d 0 R\r\n", l->index);
         }

--- a/pdfgen.c
+++ b/pdfgen.c
@@ -98,8 +98,15 @@
 #include <BaseTsd.h>
 typedef SSIZE_T ssize_t;
 #else
+
+#ifndef _POSIX_SOURCE
 #define _POSIX_SOURCE     /* For localtime_r */
+#endif
+
+#ifndef _XOPEN_SOURCE
 #define _XOPEN_SOURCE 600 /* for M_SQRT2 */
+#endif
+
 #include <sys/types.h>    /* for ssize_t */
 #endif
 

--- a/pdfgen.c
+++ b/pdfgen.c
@@ -298,7 +298,7 @@ static int flexarray_set(struct flexarray *flex, int index, void *data)
         return -EINVAL;
     if (bin >= flex->bin_count) {
         void ***bins = (void ***)realloc(flex->bins, (flex->bin_count + 1) *
-                                                         sizeof(flex->bins));
+                                                         sizeof(*flex->bins));
         if (!bins)
             return -ENOMEM;
         flex->bin_count++;

--- a/pdfgen.c
+++ b/pdfgen.c
@@ -247,7 +247,7 @@ static int bin_offset[] = {
 static inline int flexarray_get_bin(const struct flexarray *flex, int index)
 {
     (void)flex;
-    for (int i = 0; i < ARRAY_SIZE(bin_offset); i++)
+    for (size_t i = 0; i < ARRAY_SIZE(bin_offset); i++)
         if (index < bin_offset[i])
             return i - 1;
     return -1;
@@ -257,7 +257,7 @@ static inline int flexarray_get_bin_size(const struct flexarray *flex,
                                          int bin)
 {
     (void)flex;
-    if (bin >= ARRAY_SIZE(bin_offset) - 1)
+    if (bin >= (int)ARRAY_SIZE(bin_offset) - 1)
         return -1;
     int next = bin_offset[bin + 1];
     return next - bin_offset[bin];
@@ -290,14 +290,14 @@ static int flexarray_set(struct flexarray *flex, int index, void *data)
     if (bin < 0)
         return -EINVAL;
     if (bin >= flex->bin_count) {
-        void *bins =
-            realloc(flex->bins, (flex->bin_count + 1) * sizeof(flex->bins));
+        void ***bins =
+            (void ***)realloc(flex->bins, (flex->bin_count + 1) * sizeof(flex->bins));
         if (!bins)
             return -ENOMEM;
         flex->bin_count++;
         flex->bins = bins;
         flex->bins[flex->bin_count - 1] =
-            calloc(flexarray_get_bin_size(flex, flex->bin_count - 1),
+            (void **)calloc(flexarray_get_bin_size(flex, flex->bin_count - 1),
                    sizeof(void *));
         if (!flex->bins[flex->bin_count - 1]) {
             flex->bin_count--;
@@ -359,7 +359,7 @@ static ssize_t dstr_ensure(struct dstr *str, size_t len)
         bool have_stack_data = !str->data;
 
         new_len = len + 4096;
-        new_data = realloc(str->data, new_len);
+        new_data = (char *)realloc(str->data, new_len);
         if (!new_data)
             return -ENOMEM;
         // If we move beyond the on-stack buffer, copy the old data out
@@ -373,7 +373,7 @@ static ssize_t dstr_ensure(struct dstr *str, size_t len)
 
 #ifndef SKIP_ATTRIBUTE
 static int dstr_printf(struct dstr *str, const char *fmt, ...)
-    __attribute__((format(gnu_printf, 2, 3)));
+    __attribute__((format(printf, 2, 3)));
 #endif
 static int dstr_printf(struct dstr *str, const char *fmt, ...)
 {
@@ -425,7 +425,7 @@ static void dstr_free(struct dstr *str)
 
 #ifndef SKIP_ATTRIBUTE
 static int pdf_set_err(struct pdf_doc *doc, int errval, const char *buffer,
-                       ...) __attribute__((format(gnu_printf, 3, 4)));
+                       ...) __attribute__((format(printf, 3, 4)));
 #endif
 static int pdf_set_err(struct pdf_doc *doc, int errval, const char *buffer,
                        ...)
@@ -442,8 +442,8 @@ static int pdf_set_err(struct pdf_doc *doc, int errval, const char *buffer,
         return errval;
     }
 
-    if (len >= sizeof(doc->errstr) - 1)
-        len = sizeof(doc->errstr) - 1;
+    if (len >= (int)(sizeof(doc->errstr) - 1))
+        len = (int)(sizeof(doc->errstr) - 1);
 
     /* Make sure we're properly terminated */
     if (len > 0 && doc->errstr[len - 1] != '\n')
@@ -475,7 +475,7 @@ void pdf_clear_err(struct pdf_doc *pdf)
 
 static struct pdf_object *pdf_get_object(const struct pdf_doc *pdf, int index)
 {
-    return flexarray_get(&pdf->objects, index);
+    return (struct pdf_object *)flexarray_get(&pdf->objects, index);
 }
 
 static int pdf_append_object(struct pdf_doc *pdf, struct pdf_object *obj)
@@ -522,7 +522,7 @@ static struct pdf_object *pdf_add_object(struct pdf_doc *pdf, int type)
 {
     struct pdf_object *obj;
 
-    obj = calloc(1, sizeof(*obj));
+    obj = (struct pdf_object *)calloc(1, sizeof(*obj));
     if (!obj) {
         pdf_set_err(pdf, -errno,
                     "Unable to allocate object %d of type %d: %s",
@@ -573,7 +573,7 @@ struct pdf_doc *pdf_create(float width, float height,
     struct pdf_doc *pdf;
     struct pdf_object *obj;
 
-    pdf = calloc(1, sizeof(*pdf));
+    pdf = (struct pdf_doc *)calloc(1, sizeof(*pdf));
     if (!pdf)
         return NULL;
     pdf->width = width;
@@ -588,7 +588,7 @@ struct pdf_doc *pdf_create(float width, float height,
         pdf_destroy(pdf);
         return NULL;
     }
-    obj->info = calloc(sizeof(*obj->info), 1);
+    obj->info = (struct pdf_info *)calloc(sizeof(*obj->info), 1);
     if (!obj->info) {
         pdf_destroy(pdf);
         return NULL;
@@ -802,7 +802,7 @@ static int pdf_save_object(struct pdf_doc *pdf, FILE *fp, int index)
         fprintf(fp, "/Contents [\r\n");
         for (int i = 0; i < flexarray_size(&object->page.children); i++) {
             struct pdf_object *child =
-                flexarray_get(&object->page.children, i);
+                (struct pdf_object *)flexarray_get(&object->page.children, i);
             fprintf(fp, "%d 0 R\r\n", child->index);
         }
         fprintf(fp, "]\r\n");
@@ -831,8 +831,8 @@ static int pdf_save_object(struct pdf_doc *pdf, FILE *fp, int index)
         int nchildren = flexarray_size(&object->bookmark.children);
         if (nchildren > 0) {
             struct pdf_object *f, *l;
-            f = flexarray_get(&object->bookmark.children, 0);
-            l = flexarray_get(&object->bookmark.children, nchildren - 1);
+            f = (struct pdf_object *)flexarray_get(&object->bookmark.children, 0);
+            l = (struct pdf_object *)flexarray_get(&object->bookmark.children, nchildren - 1);
             fprintf(fp, "/First %d 0 R\r\n", f->index);
             fprintf(fp, "/Last %d 0 R\r\n", l->index);
         }
@@ -1605,8 +1605,8 @@ int pdf_add_text_wrap(struct pdf_doc *pdf, struct pdf_object *page,
         if (output) {
             int len = end - start;
             float char_spacing = 0;
-            if (len >= sizeof(line))
-                len = sizeof(line) - 1;
+            if (len >= (int)sizeof(line))
+                len = (int)sizeof(line) - 1;
             strncpy(line, start, len);
             line[len] = '\0';
 
@@ -1938,7 +1938,7 @@ static const struct {
 
 static int find_128_encoding(char ch)
 {
-    for (int i = 0; i < ARRAY_SIZE(code_128a_encoding); i++) {
+    for (int i = 0; i < (int)ARRAY_SIZE(code_128a_encoding); i++) {
         if (code_128a_encoding[i].ch == ch)
             return i;
     }
@@ -2025,7 +2025,7 @@ static const struct {
 
 static int find_39_encoding(char ch)
 {
-    for (int i = 0; i < ARRAY_SIZE(code_39_encoding); i++) {
+    for (int i = 0; i < (int)ARRAY_SIZE(code_39_encoding); i++) {
         if (code_39_encoding[i].ch == ch) {
             return code_39_encoding[i].code;
         }
@@ -2236,7 +2236,7 @@ static pdf_object *pdf_add_raw_jpeg(struct pdf_doc *pdf,
         return NULL;
     }
 
-    jpeg_data = malloc(len);
+    jpeg_data = (uint8_t *)malloc(len);
     if (!jpeg_data) {
         pdf_set_err(pdf, -errno, "Unable to allocate: %zu", len);
         fclose(fp);
@@ -2336,7 +2336,7 @@ int pdf_add_ppm(struct pdf_doc *pdf, struct pdf_object *page, float x,
     }
 
     size = width * height * 3;
-    data = malloc(size);
+    data = (uint8_t *)malloc(size);
     if (!data) {
         fclose(fp);
         return pdf_set_err(pdf, -ENOMEM,

